XMonad/Hooks/StatusBar/PP.hs

```diff
@@ -200,17 +210,105 @@ pprWindowSet :: WorkspaceSort -> [Window] -> PP -> WindowSet -> String
 pprWindowSet sort' urgents pp s = sepBy (ppWsSep pp) . map fmt . sort' $
             map S.workspace (S.current s : S.visible s) ++ S.hidden s
   where
-    this     = S.currentTag s
-    visibles = map (S.tag . S.workspace) (S.visible s)
+    preds = ppPrinters pp ++
+        [ (isUrgent           , ppUrgent)
+        , (isCurrent'         , ppCurrent)
+        , (isVisible'         , ppVisible)
+        , (isVisibleNoWindows', liftA2 fromMaybe ppVisible ppVisibleNoWindows)
+        , (isHidden'          , ppHidden)
+        ]
 
     fmt w = printer pp (ppRename pp (S.tag w) w)
       where
-        printer | any (\x -> (== Just (S.tag w)) (S.findTag x s)) urgents = ppUrgent
-                | S.tag w == this                                         = ppCurrent
-                | S.tag w `elem` visibles && isJust (S.stack w)           = ppVisible
-                | S.tag w `elem` visibles                                 = liftA2 fromMaybe ppVisible ppVisibleNoWindows
-                | isJust (S.stack w)                                      = ppHidden
-                | otherwise                                               = ppHiddenNoWindows
+        printer | Just (_, f) <- find (\pf ->  fst pf urgents s w) preds = f
+                | otherwise                                              = ppHiddenNoWindows
+
+-- $predicates
+-- Using 'WorkspacePredicate' and 'WorkspaceFormatter' with
+-- 'ppPrinters' can allow users to define new workspace types beyond
+-- current, urgent, and the likes that are defined in 'PP' fields. The
+-- 'WorkspaceFormatter' corresponding to the first
+-- 'WorkspacePredicate' that returns 'True' in 'ppPrinters' is used to
+-- format the 'WorkspaceId'. If nothing matches, the default workspace
+-- formatters are used.
+
+type WorkspaceFormatter = PP -> WorkspaceId -> String
+type WorkspacePredicate = [Window] -- ^ Urgent windows
+        -> WindowSet
+        -> WindowSpace
+        -> Bool
+
+-- | Predicate for urgent workspaces.
+isUrgent :: WorkspacePredicate
+isUrgent urgents s w = any (\x -> (== Just (S.tag w)) (S.findTag x s)) urgents
+
+-- | Predicate for the current workspace. Caution: assumes default
+-- precedence is respected.
```

Comment:
It may not be that clear _which_ default precedence it meant here

Reply:
https://github.com/xmonad/xmonad-contrib/pull/605/commits/55adca6f430dc85d189ced2bb6e91b4a1b00279f ?

Reply:
:+1: